targets/simple_switch/externs/sym_crypto/crypto.cpp

```diff
@@ -0,0 +1,618 @@
+      K.reserve(secure_association_key.size());
+      K.insert(K.cend(), secure_association_key.cbegin(),
+               secure_association_key.cend());
+
+      // std::cout << "[p4sec] K size " << K.size() <<  std::endl;
+      // hexdump((char*)&K[0], K.size());
+
+      // 12 byte IV
+      std::vector<unsigned char> IV;
+      IV.reserve(secure_channel_identifier.size() + packet_number.size());
+      // The 64 most significant bits of the 96-bit IV are the octets of
+      // the SCI, encoded as a binary number (9.1).
+      IV.insert(IV.cend(), secure_channel_identifier.cbegin(),
+                secure_channel_identifier.cend());
+      // The 32 least significant bits of the 96-bit IV are the octets of
+      // the PN, encoded as a binary number
+      IV.insert(IV.cend(), packet_number.cbegin(), packet_number.cend());
+
+      // std::cout << "[p4sec] IV size " << IV.size() <<  std::endl;
+      // hexdump((char*)&IV[0], IV.size());
+
+
+      // A is the Destination MAC Address, Source MAC Address, and the
+      // octets of the SecTAG concatenated in that order
+      std::vector<unsigned char> A;
+      A.reserve(destination_mac_address.size() + source_mac_address.size()
+                + security_tag.size());
+      A.insert(A.cend(), destination_mac_address.cbegin(),
+               destination_mac_address.cend());
+      A.insert(A.cend(), source_mac_address.cbegin(),
+               source_mac_address.cend());
+      A.insert(A.cend(), security_tag.cbegin(), security_tag.cend());
+
+      // P is the octets of the User Data
+      std::vector<unsigned char> P;
+      P.reserve(secure_data.size());
+      P.insert(P.cend(), secure_data.cbegin(), secure_data.cend());
+
+      out_user_data.resize(P.size(), '\0');
+
+      int actual_size = 0, final_size = 0;
+      EVP_CIPHER_CTX *d_ctx = EVP_CIPHER_CTX_new();
+      EVP_DecryptInit(d_ctx, EVP_aes_128_gcm(), &K[0], &IV[0]);
+
+      // https://www.openssl.org/docs/man1.0.2/crypto/EVP_get_cipherbynid.html#GCM_Mode
+      // To specify any additional authenticated data (AAD) a call to
+      // EVP_CipherUpdate(), EVP_EncryptUpdate() or EVP_DecryptUpdate() should
+      // be made with the output parameter out set to NULL
+      EVP_DecryptUpdate(d_ctx, NULL, &actual_size, &A[0], A.size());
+
+      EVP_DecryptUpdate(d_ctx, &out_user_data[0], &actual_size,
+                        &P[0], P.size() );
+      EVP_CIPHER_CTX_ctrl(d_ctx, EVP_CTRL_GCM_SET_TAG, 16,
+                          &integrity_check_value[0]);
+      int result = EVP_DecryptFinal(d_ctx, &out_user_data[actual_size],
+                                    &final_size);
+
+      if (result == 1) {
+        // valid result
+      } else {
+        // decryption failed
+        // -> abprt/drop packet?
+      }
+
+      std::cout << "result of decryption: " << result << std::endl;
+
+
+      EVP_CIPHER_CTX_free(d_ctx);
+
+      return result;
+  }
+
+  void hexdump(char *addr, int len) {
```

Comment:
it makes no sense to cast the argument to a `char *` at the callsite, only to cast it back to an `unsigned char *` at the beginning of this function.